src/common/s3.connection.cpp

```diff
@@ -0,0 +1,569 @@
+        return (bool)response;
+    } catch (const std::exception& e) {
+        LOGE("Failed to create multipart object %s in bucket %s: %s",
+             object_name.c_str(),
+             bucket_name.c_str(),
+             e.what());
+    }
+
+    return false;
+}
+
+bool
+common::S3Connection::upload_multipart_object_part(
+  const std::string& bucket_name,
+  const std::string& object_name,
+  const std::string& upload_id,
+  const uint8_t* data,
+  size_t nbytes,
+  size_t part_number,
+  std::string& etag)
+{
+    if (bucket_name.empty() || object_name.empty() || 0 == nbytes) {
+        return false;
+    }
+
+    TRACE("Uploading multipart object part %zu for object %s in bucket %s",
+          part_number,
+          object_name.c_str(),
+          bucket_name.c_str());
+
+    std::string_view sv(reinterpret_cast<const char*>(data), nbytes);
+
+    minio::s3::UploadPartArgs args;
+    args.bucket = bucket_name;
+    args.object = object_name;
+    args.part_number = part_number;
+    args.upload_id = upload_id;
+    args.data = sv;
+
+    try {
+        minio::s3::UploadPartResponse response = client_->UploadPart(args);
+        etag = response.etag;
+
+        return (bool)response;
+    } catch (const std::exception& e) {
+        LOGE("Failed to upload multipart object part %zu for object %s in "
+             "bucket %s: %s",
+             part_number,
+             object_name.c_str(),
+             bucket_name.c_str(),
+             e.what());
+    }
+
+    return false;
+}
+
+bool
+common::S3Connection::complete_multipart_object(
+  const std::string& bucket_name,
+  const std::string& object_name,
+  const std::string& upload_id,
+  const std::list<minio::s3::Part>& parts)
+{
+    if (bucket_name.empty() || object_name.empty()) {
+        return false;
+    }
+
+    TRACE("Completing multipart object %s in bucket %s",
+          object_name.c_str(),
+          bucket_name.c_str());
+    minio::s3::CompleteMultipartUploadArgs args;
+    args.bucket = bucket_name;
+    args.object = object_name;
+    args.upload_id = upload_id;
+    args.parts = parts;
+
+    try {
+        minio::s3::CompleteMultipartUploadResponse response =
+          client_->CompleteMultipartUpload(args);
+
+        return (bool)response;
+    } catch (const std::exception& e) {
+        LOGE("Failed to complete multipart object %s in bucket %s: %s",
+             object_name.c_str(),
+             bucket_name.c_str(),
+             e.what());
+    }
+
+    return false;
+}
+
+common::S3ConnectionPool::S3ConnectionPool(size_t n_connections,
+                                           const std::string& endpoint,
+                                           const std::string& access_key_id,
+                                           const std::string& secret_access_key)
+  : is_accepting_connections_{ true }
+{
+    CHECK(n_connections > 0);
+
+    for (auto i = 0; i < n_connections; ++i) {
+        connections_.push_back(std::make_unique<S3Connection>(
```

Comment:
I’m a little worried that we don’t check if the connection was successful immediately after it’s created. Based on a quick skim of the docs (and I could be wrong because I haven’t actually tested this), any issues with the connection will be revealed when we try to make a request.

Is it possible then for us to have a client in the pool, or the entire pool, containing clients that will always raise errors? And what if a specific connection fails — shouldn’t we add logic to ensure that it is permanently removed from the pool? If we do, what happens if the pool becomes empty? Would we have a bunch of threads just waiting for a connection to appear?

Reply:
Yeah, this is a good point. You're correct that you won't find out about errors in the connection til you actually try an operation. That's how I found out about the `RequestTimeTooSkewed` [issue](https://github.com/minio/minio-cpp/issues/150).